### PR TITLE
Fix VAAPI encoding on 10-bit hevc. Update QSV encoding and scaling to...

### PR DESCRIPTION
**Pull request by cryptobank**
… take advantage of 1) scale_qsv and 2) overlay_qsv and 3) deinterlace_qsv.

Changes
scale_vaapi=format=nv12, numerous QSV changes (hardware scaling, hardware overlay, hardware deinterlace). There's currently an issue with ffmpeg QSV where it's not possible to change HLS time(stuck at 10 seconds) and force key frames (breaks HLS) therefore i've added a check in DynamicHlsService.cs for this. It means that seeking is somewhat broken with QSV but otherwise it seems to work and is speedy (quicker than VAAPI in most cases)

Issues
At least for me this fixes numerous HEVC 10 bit transcode issues that have previously been reported. #2046 #1901 #1527 to name a few.

Update:
I've rebased against jellyfin/master and fixed formatting as requested.

Please review. I'm no C# hacker, I haven't programmed since my old C++ days 10 years ago so I'm happy to take criticism.

### CODE REVIEW

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 1]
```diff
@@ -945,7 +945,12 @@ protected override string GetVideoArguments(StreamState state, EncodingOptions e
 
                 var hasGraphicalSubs = state.SubtitleStream != null && !state.SubtitleStream.IsTextSubtitleStream && state.SubtitleDeliveryMethod == SubtitleDeliveryMethod.Encode;
 
-                args += " " + EncodingHelper.GetVideoQualityParam(state, codec, encodingOptions, GetDefaultEncoderPreset()) + keyFrameArg;
+                args += " " + EncodingHelper.GetVideoQualityParam(state, codec, encodingOptions, GetDefaultEncoderPreset());
+
+                // Unable to force key frames to h264_qsv transcode
+                if (codec != "h264_qsv") {
+                    args += " " + keyFrameArg; 
+                }
```
Comment (JustAMan) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: there should be a debug logging saying "cannot force keyframes" or something similar so we won't lose it

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 2]
```diff
@@ -1536,8 +1558,8 @@ public string GetGraphicalSubtitleParam(EncodingJobInfo state, EncodingOptions o
                     {
                         outputSizeParam = "," + outputSizeParam.Substring(index);
                     }
-                }
-                else
+                } 
+                else
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: trailing spaces

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 3]
```diff
@@ -1570,7 +1593,15 @@ public string GetGraphicalSubtitleParam(EncodingJobInfo state, EncodingOptions o
...
 
-                videoSizeParam += ":force_original_aspect_ratio=decrease";
+                if (string.Equals(outputVideoCodec, "h264_qsv", StringComparison.OrdinalIgnoreCase))
+                {
+                    //For QSV, feed it into hardware encoder
+                    videoSizeParam += ",hwupload=extra_hw_frames=64";
+                } 
+                else 
+                {
+                    videoSizeParam += ":force_original_aspect_ratio=decrease";
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: this changes the non-`qsv` behaviour, why do you want to stuff it in QSV-related PR?

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 4]
```diff
@@ -1914,6 +1970,22 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
...
+            {
+                filters.Add("format=nv12|qsv");
+                filters.Add("hwupload=extra_hw_frames=64");
+            }
+
+            var inputWidth = videoStream?.Width;
+            var inputHeight = videoStream?.Height;
+            var threeDFormat = state.MediaSource.Video3DFormat;
+
+            filters.AddRange(GetScalingFilters(inputWidth, inputHeight, threeDFormat, videoDecoder, outputVideoCodec, request.Width, request.Height, request.MaxWidth, request.MaxHeight));
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: why did you move these?

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 5]
```diff
@@ -2134,10 +2199,9 @@ public string GetInputModifier(EncodingJobInfo state, EncodingOptions encodingOp
             }
 
             var videoDecoder = GetHardwareAcceleratedVideoDecoder(state, encodingOptions);
+
             if (!string.IsNullOrEmpty(videoDecoder))
             {
-                inputModifier += " " + videoDecoder;
-
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: why remove?

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs (hunk 1 above)
Comment (cryptobank) [reply MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: How do you log in jellyfin?

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 2 above)
Comment (cryptobank) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: Fixed

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 3 above)
Comment (cryptobank) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: force_original_aspect_ratio=decrease was the original code. i've now removed this entirely because it actually causes the bitmap subtitles to appear off-centre.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 4 above)
Comment (cryptobank) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: fixed. it was moved while working out what  the code does and i forgot to move it back afterwards.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 5 above)
Comment (cryptobank) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: fixed. same as above.

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs (hunk 1 above)
Comment (cryptobank) [reply MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: Worked it out for myself. Done.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 6]
```diff
@@ -496,6 +517,7 @@ public string GetInputArgument(EncodingJobInfo state, EncodingOptions encodingOp
                     }
                 }
 
+
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]:
```suggestion
```

**Comment by cryptobank** [comment]
@JustAMan who else is a reviewer?

**Comment by richardbrodie** [comment]
Is there some process to follow for requesting a review?

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 7]
```diff
@@ -945,7 +945,16 @@ protected override string GetVideoArguments(StreamState state, EncodingOptions e
 
                 var hasGraphicalSubs = state.SubtitleStream != null && !state.SubtitleStream.IsTextSubtitleStream && state.SubtitleDeliveryMethod == SubtitleDeliveryMethod.Encode;
 
-                args += " " + EncodingHelper.GetVideoQualityParam(state, codec, encodingOptions, GetDefaultEncoderPreset()) + keyFrameArg;
+                args += " " + EncodingHelper.GetVideoQualityParam(state, codec, encodingOptions, GetDefaultEncoderPreset());
+
+                // Unable to force key frames to h264_qsv transcode
+                if (string.Equals(codec, "h264_qsv", StringComparison.OrdinalIgnoreCase)) {
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]:
```suggestion
                if (string.Equals(codec, "h264_qsv", StringComparison.OrdinalIgnoreCase))
                {
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 8]
```diff
@@ -467,6 +467,27 @@ public string GetInputArgument(EncodingJobInfo state, EncodingOptions encodingOp
                     .Append(' ');
             }
 
+            if (state.IsVideoRequest 
+                && string.Equals(encodingOptions.HardwareAccelerationType, "qsv", StringComparison.OrdinalIgnoreCase))
+            {
+                var videoDecoder = GetHardwareAcceleratedVideoDecoder(state, encodingOptions);
+                var outputVideoCodec = GetVideoEncoder(state, encodingOptions);
+
+                if (encodingOptions.EnableHardwareEncoding && outputVideoCodec.Contains("qsv"))
```
Comment (Bond-009) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]:
```suggestion
                if (encodingOptions.EnableHardwareEncoding && outputVideoCodec.Contains("qsv", StringComparison.OrdinalIgnoreCase))
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 9]
```diff
@@ -467,6 +467,27 @@ public string GetInputArgument(EncodingJobInfo state, EncodingOptions encodingOp
...
+                {
+                    if (!string.IsNullOrEmpty(videoDecoder) && videoDecoder.Contains("qsv"))
```
Comment (Bond-009) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]:
```suggestion
                    if (!string.IsNullOrEmpty(videoDecoder) && videoDecoder.Contains("qsv", StringComparison.OrdinalIgnoreCase))
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 10]
```diff
@@ -1581,9 +1607,31 @@ public string GetGraphicalSubtitleParam(EncodingJobInfo state, EncodingOptions o
...
+            var retStr = " -filter_complex \"[{0}:{1}]{4}[sub];[0:{2}][sub]overlay{3}\"";
+
+            if (string.Equals(outputVideoCodec, "h264_qsv", StringComparison.OrdinalIgnoreCase)) 
+            {
+                /*
+                    QSV in FFMpeg can now setup hardware overlay for transcodes.
+                    For software decoding and hardware encoding option, frames must be hwuploaded into hardware
+                    with fixed frame size. 
+                */
+                if (!string.IsNullOrEmpty(videoDecoder) && videoDecoder.Contains("qsv"))
```
Comment (Bond-009) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]:
```suggestion
                if (!string.IsNullOrEmpty(videoDecoder) && videoDecoder.Contains("qsv", StringComparison.OrdinalIgnoreCase))
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 11]
```diff
@@ -1914,10 +1965,26 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
                 filters.Add("hwupload");
             }
 
-            if (state.DeInterlace("h264", true)
-                && string.Equals(outputVideoCodec, "h264_vaapi", StringComparison.OrdinalIgnoreCase))
+            var videoDecoder = GetHardwareAcceleratedVideoDecoder(state, options);
+
+            // If we are software decoding, and hardware encoding	  
+            if (string.Equals(outputVideoCodec, "h264_qsv", StringComparison.OrdinalIgnoreCase) 
+                && (string.IsNullOrEmpty(videoDecoder) || !videoDecoder.Contains("qsv")))
```
Comment (Bond-009) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]:
```suggestion
                && (string.IsNullOrEmpty(videoDecoder) || !videoDecoder.Contains("qsv", StringComparison.OrdinalIgnoreCase)))
```
Maybe add braces to make the order more obvious.

**Comment by JustAMan** [comment]
> @JustAMan who else is a reviewer?

Whoever from @jellyfin/backend feels up to it. No certain procedure, as we're all volunteers :)

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 12]
```diff
@@ -1647,31 +1695,34 @@ public string GetGraphicalSubtitleParam(EncodingJobInfo state, EncodingOptions o
...
                 var outputWidth = width.Value;
                 var outputHeight = height.Value;
+                var vaapi_or_qsv = string.Equals(videoEncoder, "h264_qsv", StringComparison.OrdinalIgnoreCase) ? "qsv" : "vaapi";
 
                 if (!videoWidth.HasValue
                     || outputWidth != videoWidth.Value
                     || !videoHeight.HasValue
                     || outputHeight != videoHeight.Value)
                 {
+                    //Force nv12 pixel format to enable 10-bit to 8-bit colour conversion.
```
Comment (Bond-009) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]:
```suggestion
                    // Force nv12 pixel format to enable 10-bit to 8-bit colour conversion.
```
